domain/credential/state/credential.go

```diff
@@ -0,0 +1,489 @@
+		return nil, errors.Trace(err)
+	}
+	result := make(map[string]cloud.Credential)
+	for _, cred := range creds {
+		result[cred.Label] = cred
+	}
+	return result, nil
+}
+
+// CloudCredential returns the cloud credential for the given tag.
+func (st *State) CloudCredential(ctx context.Context, name, cloudName, owner string) (cloud.Credential, error) {
+	db, err := st.DB()
+	if err != nil {
+		return cloud.Credential{}, errors.Trace(err)
+	}
+
+	var creds []cloud.Credential
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		var err error
+		creds, err = st.loadCloudCredentials(ctx, tx, name, cloudName, owner)
+		return errors.Trace(err)
+	})
+	if err != nil {
+		return cloud.Credential{}, errors.Trace(err)
+	}
+	if len(creds) == 0 {
+		return cloud.Credential{}, fmt.Errorf("credential %q for cloud %q owned by %q %w", name, cloudName, owner, errors.NotFound)
+	}
+	if len(creds) > 1 {
+		return cloud.Credential{}, errors.Errorf("expected 1 credential, got %d", len(creds))
+	}
+	return creds[0], errors.Trace(err)
+}
+
+type credentialInfo struct {
+	id            string
+	name          string
+	revoked       bool
+	invalid       bool
+	invalidReason string
+	authType      string
+	cloudName     string
+	owner         string
+}
+
+type credentialAttrs map[string]string
+
+func (st *State) loadCloudCredentials(ctx context.Context, tx *sql.Tx, name, cloudName, owner string) ([]cloud.Credential, error) {
+	// First load the basic credential info.
+	credQuery := `
+SELECT cloud_credential.uuid, cloud_credential.name,
+       cloud_credential.revoked, cloud_credential.invalid, 
+       cloud_credential.invalid_reason, 
+       cloud_credential.owner_uuid AS owner,
+       auth_type.type AS auth_type, cloud.name AS cloud_name,
+       cloud_credential_attributes.key, cloud_credential_attributes.value
+FROM   cloud_credential
+       INNER JOIN auth_type 
+             ON cloud_credential.auth_type_id = auth_type.id
+       INNER JOIN cloud 
+             ON cloud_credential.cloud_uuid = cloud.uuid
+       LEFT JOIN cloud_credential_attributes
+            ON cloud_credential_attributes.cloud_credential_uuid = cloud_credential.uuid
+`
+
+	var (
+		args        []any
+		filterTerms []string
+	)
+	if name != "" {
+		filterTerms = append(filterTerms, "cloud_credential.name = ?")
+		args = append(args, name)
+	}
+	if cloudName != "" {
+		filterTerms = append(filterTerms, "cloud.name = ?")
+		args = append(args, cloudName)
+	}
+	if owner != "" {
+		filterTerms = append(filterTerms, "cloud_credential.owner_uuid = ?")
+		args = append(args, owner)
+	}
+	if len(filterTerms) > 0 {
+		credQuery = credQuery + "\nWHERE " + strings.Join(filterTerms, " AND ")
+	}
+
+	rows, err := tx.QueryContext(ctx, credQuery, args...)
+	if err != nil && err != sql.ErrNoRows {
+		return nil, errors.Trace(err)
+	}
+	defer func() { _ = rows.Close() }()
+
+	var result []cloud.Credential
+	recordResult := func(info *credentialInfo, attrs credentialAttrs) {
```

Comment:
We don't want to do any of this. Have a look at the state for external controllers.
- Create a row type for the query and populate it with SQLair.
- Create a type for the slice of that row type and add a method that converts it to a normalised domain type.

Reply:
I'll rework it as requested, but it seems the gist of the logic is very similar - you still need to iterate over all the rows and de-dupe the parent records and then aggregate and apply the referenced one-to-many or child records. I guess the example referenced doing this:
```
	for _, controller := range e {
		uniqueControllers[controller.ID] = crossmodel.ControllerInfo{
			ControllerTag: names.NewControllerTag(controller.ID),
			CACert:        controller.CACert,
			Alias:         controller.Alias.String,
		}
		...
```
seems somewhat inefficient having to create a new struct and update a map for each and every row in the result set. As opposed to the technique in this PR which does it only when the uuid changes when iterating the rows. (actually, this PR is missing an ORDER BY). 
Given the algorithm is similar in both approaches, to what extent is this just a matter of preference. I prefer the method in this PR, YMMV. But it's better to be consistent so if the external controllers approach is mandated, then I will live with it.

Reply:
done